### PR TITLE
Add a fuzz test for the synthetic node impl

### PR DESCRIPTION
**Pull request by ljedrz**
This PR adds a `decoder` fuzz target that can be used with `cargo-fuzz` (`cargo +nightly fuzz run decoder`) and a few patches that fix issues that I've found with it, now that the synthetic node can be used outside of tests.

Cc @tuommaki

### CODE REVIEW

File: src/protocol/payload/codec.rs [hunk 1]
```diff
@@ -32,8 +32,6 @@ impl<T: Codec> Codec for Vec<T> {
         Self: Sized,
     {
         let length = *VarInt::decode(bytes)?;
-        (0..length)
-            .map(|_| T::decode(bytes))
-            .collect::<io::Result<Self>>()
+        (0..length).map(|_| T::decode(bytes)).collect()
```
Comment (ljedrz) [inline src/protocol/payload/codec.rs:35]: this one is just a drive-by cleanup

File: src/protocol/payload/addr.rs [hunk 2]
```diff
@@ -86,6 +86,10 @@ impl NetworkAddr {
     pub(super) fn decode_without_timestamp<B: Buf>(bytes: &mut B) -> io::Result<Self> {
         let services = u64::from_le_bytes(read_n_bytes(bytes)?);
 
+        if bytes.remaining() < 16 {
```
Comment (niklaslong) [inline src/protocol/payload/addr.rs:89]: We could consider using `read_n_bytes` everywhere instead, since it checks for the length. Not sure if it would be slower than the built in `get_*` methods.

**Review by niklaslong — APPROVED** [review]
LGTM 👍

File: src/protocol/payload/addr.rs (hunk 2 above)
Comment (ljedrz) [reply src/protocol/payload/addr.rs:89]: yeah, I had the same thoughts; we might want to align this at one point in time, but I'm ok with both options
